postgresql/resource_postgresql_function.go

```diff
@@ -0,0 +1,348 @@
+			},
+		},
+	}
+}
+
+func resourcePostgreSQLFunctionCreate(db *DBConnection, d *schema.ResourceData) error {
+	if !db.featureSupported(featureFunction) {
+		return fmt.Errorf(
+			"postgresql_function resource is not supported for this Postgres version (%s)",
+			db.version,
+		)
+	}
+
+	if err := createFunction(db, d, false); err != nil {
+		return err
+	}
+
+	return resourcePostgreSQLFunctionReadImpl(db, d)
+}
+
+func resourcePostgreSQLFunctionExists(db *DBConnection, d *schema.ResourceData) (bool, error) {
+	if !db.featureSupported(featureFunction) {
+		return false, fmt.Errorf(
+			"postgresql_function resource is not supported for this Postgres version (%s)",
+			db.version,
+		)
+	}
+
+	signature := getFunctionSignature(d)
+	functionExists := false
+
+	txn, err := startTransaction(db.client, "")
+	if err != nil {
+		return false, err
+	}
+	defer deferredRollback(txn)
+
+	query := fmt.Sprintf("SELECT to_regprocedure('%s') IS NOT NULL", signature)
+	err = txn.QueryRow(query).Scan(&functionExists)
+	return functionExists, err
+}
+
+func resourcePostgreSQLFunctionRead(db *DBConnection, d *schema.ResourceData) error {
+	if !db.featureSupported(featureFunction) {
+		return fmt.Errorf(
+			"postgresql_function resource is not supported for this Postgres version (%s)",
+			db.version,
+		)
+	}
+
+	return resourcePostgreSQLFunctionReadImpl(db, d)
+}
+
+func resourcePostgreSQLFunctionReadImpl(db *DBConnection, d *schema.ResourceData) error {
+	signature := getFunctionSignature(d)
+
+	txn, err := startTransaction(db.client, "")
+	if err != nil {
+		return err
+	}
+	defer deferredRollback(txn)
+
+	var funcSchema, funcName string
+
+	query := `SELECT n.nspname, p.proname ` +
+		`FROM pg_proc p ` +
+		`LEFT JOIN pg_namespace n ON p.pronamespace = n.oid ` +
+		`WHERE p.oid = to_regprocedure($1)`
+	err = txn.QueryRow(query, signature).Scan(&funcSchema, &funcName)
+	switch {
+	case err == sql.ErrNoRows:
+		log.Printf("[WARN] PostgreSQL function: %s", signature)
+		d.SetId("")
+		return nil
+	case err != nil:
+		return fmt.Errorf("Error reading function: %w", err)
+	}
+
+	d.Set(funcNameAttr, funcName)
+	d.Set(funcSchemaAttr, funcSchema)
+	d.SetId(signature)
+
+	return nil
+}
+
+func resourcePostgreSQLFunctionDelete(db *DBConnection, d *schema.ResourceData) error {
+	if !db.featureSupported(featureFunction) {
+		return fmt.Errorf(
+			"postgresql_function resource is not supported for this Postgres version (%s)",
+			db.version,
+		)
+	}
+
+	signature := getFunctionSignature(d)
+
+	txn, err := startTransaction(db.client, "")
+	if err != nil {
+		return err
+	}
+	defer deferredRollback(txn)
+
+	dropMode := "RESTRICT"
+	if v, ok := d.GetOk(funcDropCascadeAttr); ok && v.(bool) {
+		dropMode = "CASCADE"
+	}
+
+	sql := fmt.Sprintf("DROP FUNCTION IF EXISTS %s %s", signature, dropMode)
+	if _, err := txn.Exec(sql); err != nil {
+		return err
+	}
+
+	if err = txn.Commit(); err != nil {
+		return fmt.Errorf("Error deleting function: %w", err)
+	}
```

Comment:
```suggestion
	dropMode := "RESTRICT"
	if v, ok := d.GetOk(funcDropCascadeAttr); ok && v.(bool) {
		dropMode = "CASCADE"
	}

	sql := fmt.Sprintf("DROP FUNCTION IF EXISTS %s %s", signature, dropMode)
	if _, err := db.Exec(sql); err != nil {
                return fmt.Errorf("Error deleting function: %w", err)
	}
```

To simplify the code, if you execute only one SQL query within a transaction, you don't need to create it
(Same comment for all the functions)

Reply:
Done